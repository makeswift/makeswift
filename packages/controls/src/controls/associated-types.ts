export type AssociatedTypes<T> = T extends {
  __associatedTypes(types: infer Types): void
}
  ? Types extends Record<string, unknown>
    ? Types
    : never
  : never

export type AssociatedType<
  T,
  K extends keyof AssociatedTypes<T>,
> = AssociatedTypes<T>[K]

export type ControlType<D> = AssociatedType<D, 'ControlType'>
export type ConfigType<D> = AssociatedType<D, 'Config'>
export type DataType<D> = AssociatedType<D, 'DataType'>
export type ValueType<D> = AssociatedType<D, 'ValueType'>
export type ResolvedValueType<D> = AssociatedType<D, 'ResolvedValueType'>
export type InstanceType<D> = AssociatedType<D, 'InstanceType'>
