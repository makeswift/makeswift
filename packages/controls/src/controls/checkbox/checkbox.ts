import { match, P } from 'ts-pattern'
import { z } from 'zod'

import { safeParse, type ParseResult } from '../../lib/zod'

import { ControlDataTypeKey } from '../../common'
import { type CopyContext } from '../../context'
import { type DeserializedRecord } from '../../serialization'

import {
  ControlDefinition,
  type Resolvable,
  type SchemaType,
} from '../definition'
import { DefaultControlInstance, type SendMessage } from '../instance'
import { ControlDefinitionVisitor } from '../visitor'

type Config = z.infer<typeof Definition.schema.relaxed.config>

type SchemaByDefaultValue<D extends Config['defaultValue']> =
  undefined extends D
    ? typeof Definition.schema.relaxed
    : typeof Definition.schema.strict

type Schema<C extends Config> = SchemaByDefaultValue<C['defaultValue']>
type DataType<C extends Config> = z.infer<Schema<C>['data']>
type ValueType<C extends Config> = z.infer<Schema<C>['value']>
type ResolvedValueType<C extends Config> = z.infer<Schema<C>['resolvedValue']>

type ReturnedSchemaType<C extends Config> = {
  definition: typeof Definition.schema.relaxed.definition
  type: typeof Definition.schema.relaxed.type
  data: SchemaType<DataType<C>>
  value: SchemaType<ValueType<C>>
  resolvedValue: SchemaType<ResolvedValueType<C>>
}

class Definition<C extends Config> extends ControlDefinition<
  typeof Definition.type,
  C,
  DataType<C>,
  ValueType<C>,
  ResolvedValueType<C>
> {
  private static readonly v1DataType = 'checkbox::v1' as const
  private static readonly dataSignature = {
    v1: { [ControlDataTypeKey]: this.v1DataType },
  } as const

  static readonly type = 'makeswift::controls::checkbox' as const

  static get schema() {
    const version = z.literal(1).optional()
    const versionedData = z.object({
      [ControlDataTypeKey]: z.literal(this.v1DataType),
      value: z.boolean(),
    })

    const schemas = <V, D>(value: SchemaType<V>, data: SchemaType<D>) => {
      const type = z.literal(this.type)

      const config = z.object({
        label: z.string().optional(),
        description: z.string().optional(),
        defaultValue: value,
      })

      const definition = z.object({
        type,
        config,
        version,
      })

      return {
        type,
        value,
        resolvedValue: value,
        data,
        versionedData,
        config,
        version,
        definition,
      }
    }

    return {
      version,
      relaxed: schemas(
        z.boolean().optional(),
        z.union([z.boolean(), versionedData, z.undefined()]),
      ),
      strict: schemas(z.boolean(), z.union([z.boolean(), versionedData])),
    }
  }

  static deserialize(data: DeserializedRecord): CheckboxDefinition {
    if (data.type !== Definition.type)
      throw new Error(
        `Checkbox: expected '${Definition.type}', got '${data.type}'`,
      )

    const { config, version } = Definition.schema.relaxed.definition.parse(data)
    return new CheckboxDefinition(config, version)
  }

  constructor(
    config: C,
    readonly version: z.infer<typeof Definition.schema.version>,
  ) {
    super(config)
  }

  get controlType() {
    return Definition.type
  }

  get schema(): ReturnedSchemaType<C> {
    return Definition.schema.relaxed
  }

  get dataSchema() {
    return (
      (this.config.defaultValue === undefined
        ? Definition.schema.relaxed
        : Definition.schema.strict) as Schema<C>
    ).data
  }

  safeParse(data: unknown | undefined): ParseResult<DataType<C> | undefined> {
    return safeParse(this.dataSchema, data)
  }

  fromData(data: DataType<C> | undefined): ValueType<C> | undefined {
    const inputSchema = this.dataSchema.optional()
    return match(data satisfies z.infer<typeof inputSchema>)
      .with(Definition.dataSignature.v1, ({ value }) => value)
      .otherwise((value) => value)
  }

  toData(value: ValueType<C>): DataType<C> {
    return match({ version: this.version, value })
      .with({ version: 1, value: P.boolean }, ({ value }) => ({
        ...Definition.dataSignature.v1,
        value,
      }))
      .with({ version: 1, value: undefined }, () => undefined)
      .with({ version: undefined }, ({ value }) => value)
      .otherwise(() => value)
  }

  copyData(
    data: DataType<C> | undefined,
    _context: CopyContext,
  ): DataType<C> | undefined {
    return data
  }

  resolveValue(
    data: DataType<C> | undefined,
  ): Resolvable<ResolvedValueType<C> | undefined> {
    return {
      name: Definition.type,
      readStable: () => this.fromData(data) ?? this.config.defaultValue,
      subscribe: () => () => {},
      triggerResolve: async () => {},
    }
  }

  createInstance(sendMessage: SendMessage<any>) {
    return new DefaultControlInstance(sendMessage)
  }

  accept<R>(visitor: ControlDefinitionVisitor<R>, ...args: unknown[]): R {
    return visitor.visitCheckbox(this, ...args)
  }
}

export class CheckboxDefinition<
  C extends Config = Config,
> extends Definition<C> {}

type UserConfig<D extends Config['defaultValue']> = Config & {
  defaultValue?: D
}

type NormedConfig<D extends Config['defaultValue']> = z.infer<
  SchemaByDefaultValue<D>['config']
>

export function Checkbox<D extends Config['defaultValue']>(
  config?: UserConfig<D>,
): CheckboxDefinition<NormedConfig<D>> {
  return new CheckboxDefinition((config ?? {}) as NormedConfig<D>, 1)
}
