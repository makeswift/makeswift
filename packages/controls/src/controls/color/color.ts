import { match } from 'ts-pattern'
import { z } from 'zod'

import { StableValue } from '../../lib/stable-value'
import { safeParse, type ParseResult } from '../../lib/zod'

import { ControlDataTypeKey } from '../../common'
import {
  ContextResource,
  replaceResourceIfNeeded,
  shouldRemoveResource,
  type CopyContext,
} from '../../context'
import { ResourceSchema } from '../../resources'
import { type ResourceResolver } from '../../resources/resolver'
import {
  type DeserializedRecord,
  type SerializedRecord,
} from '../../serialization'

import {
  ControlDefinition,
  serialize,
  type Resolvable,
  type SchemaType,
} from '../definition'
import { DefaultControlInstance, type SendMessage } from '../instance'
import { ControlDefinitionVisitor } from '../visitor'

import { swatchToColorString } from './conversion'

type Config = z.infer<typeof Definition.schema.relaxed.config>

type SchemaByDefaultValue<D extends Config['defaultValue']> =
  undefined extends D
    ? typeof Definition.schema.relaxed
    : typeof Definition.schema.strict

type Schema<C extends Config> = SchemaByDefaultValue<C['defaultValue']>
type DataType<C extends Config> = z.infer<Schema<C>['data']>
type ValueType<C extends Config> = z.infer<Schema<C>['value']>
type ResolvedValueType<C extends Config> = z.infer<Schema<C>['resolvedValue']>

type ResolvedSwatchValueType<C extends Config> = z.infer<
  Schema<C>['resolvedSwatchValue']
>

type ReturnedSchemaType<C extends Config> = {
  definition: typeof Definition.schema.relaxed.definition
  type: typeof Definition.schema.relaxed.type
  data: SchemaType<DataType<C>>
  value: SchemaType<ValueType<C>>
  resolvedValue: SchemaType<ResolvedValueType<C>>
}

class Definition<C extends Config> extends ControlDefinition<
  typeof Definition.type,
  C,
  DataType<C>,
  ValueType<C>,
  ResolvedValueType<C>
> {
  private static readonly v1DataType = 'color::v1' as const
  private static readonly dataSignature = {
    v1: { [ControlDataTypeKey]: this.v1DataType },
  } as const

  static readonly type = 'makeswift::controls::color' as const

  static get schema() {
    const type = z.literal(this.type)
    const version = z.literal(1).optional()
    const value = ResourceSchema.colorData
    const resolvedSwatchValue = ResourceSchema.resolvedColorData

    const schemas = <R>(resolvedValue: SchemaType<R>) => {
      const data = value.merge(
        z.object({
          [ControlDataTypeKey]: z.literal(this.v1DataType).optional(),
        }),
      )

      const config = z.object({
        label: z.string().optional(),
        labelOrientation: z
          .union([z.literal('horizontal'), z.literal('vertical')])
          .optional(),
        description: z.string().optional(),
        defaultValue: resolvedValue,
        hideAlpha: z.boolean().optional(),
      })

      const definition = z.object({
        type,
        config,
        version,
      })

      return {
        type,
        value,
        resolvedSwatchValue,
        resolvedValue,
        data,
        config,
        version,
        definition,
      }
    }

    return {
      version,
      relaxed: schemas(z.string().optional()),
      strict: schemas(z.string()),
    }
  }

  static deserialize(data: DeserializedRecord): ColorDefinition {
    if (data.type !== Definition.type)
      throw new Error(
        `Color deserialization: expected '${Definition.type}', got '${data.type}'`,
      )

    const { config, version } = Definition.schema.relaxed.definition.parse(data)
    return new ColorDefinition(config, version)
  }

  constructor(
    config: C,
    readonly version: z.infer<typeof Definition.schema.version>,
  ) {
    super(config)
  }

  get controlType() {
    return Definition.type
  }

  get schema(): ReturnedSchemaType<C> {
    return Definition.schema.relaxed
  }

  get dataSchema() {
    return (
      (this.config.defaultValue === undefined
        ? Definition.schema.relaxed
        : Definition.schema.strict) as Schema<C>
    ).data
  }

  safeParse(data: unknown | undefined): ParseResult<DataType<C> | undefined> {
    return safeParse(this.dataSchema, data)
  }

  fromData(data: DataType<C> | undefined): ValueType<C> | undefined {
    const inputSchema = this.dataSchema.optional()
    return match(data satisfies z.infer<typeof inputSchema>)
      .with(Definition.dataSignature.v1, ({ swatchId, alpha }) => ({
        swatchId,
        alpha,
      }))
      .otherwise((value) => value)
  }

  toData(value: ValueType<C>): DataType<C> {
    return match(this.version)
      .with(1, () => ({
        ...Definition.dataSignature.v1,
        ...value,
      }))
      .with(undefined, () => value)
      .exhaustive()
  }

  copyData(
    data: DataType<C> | undefined,
    ctx: CopyContext,
  ): DataType<C> | undefined {
    if (data == null) return data

    const currentSwatchId = data.swatchId
    if (shouldRemoveResource(ContextResource.Swatch, currentSwatchId, ctx)) {
      return undefined
    }

    const inputSchema = this.dataSchema.optional()
    return match(data satisfies z.infer<typeof inputSchema>)
      .with(Definition.dataSignature.v1, (val) => ({
        ...val,
        swatchId: replaceResourceIfNeeded(
          ContextResource.Swatch,
          val.swatchId,
          ctx,
        ),
      }))
      .otherwise((val) => ({
        ...val,
        swatchId: replaceResourceIfNeeded(
          ContextResource.Swatch,
          val.swatchId,
          ctx,
        ),
      }))
  }

  resolveSwatch(
    data: DataType<C> | undefined,
    resolver: ResourceResolver,
  ): Resolvable<ResolvedSwatchValueType<C> | undefined> {
    const value = this.fromData(data)
    const swatchSub = resolver.resolveSwatch(value?.swatchId)

    const stableValue = StableValue({
      name: `${Definition.type}:swatch`,
      read: () => {
        const swatch = swatchSub.readStable()
        return swatch == null
          ? undefined
          : {
              swatch,
              alpha: value?.alpha ?? 1,
            }
      },
      deps: [swatchSub],
    })

    return {
      ...stableValue,
      triggerResolve: async (currentValue?: ResolvedSwatchValueType<C>) => {
        if (currentValue == null) {
          await swatchSub.fetch()
        }
      },
    }
  }

  resolveValue(
    data: DataType<C> | undefined,
    resolver: ResourceResolver,
  ): Resolvable<ResolvedValueType<C> | undefined> {
    const value = this.fromData(data)
    const swatch = resolver.resolveSwatch(value?.swatchId)

    const stableValue = StableValue({
      name: Definition.type,
      read: () =>
        swatchToColorString(
          swatch.readStable(),
          value?.alpha ?? 1,
          this.config.defaultValue,
        ),
      deps: [swatch],
    })

    return {
      ...stableValue,
      triggerResolve: async () => {
        if (swatch.readStable() == null) {
          await swatch.fetch()
        }
      },
    }
  }

  createInstance(sendMessage: SendMessage<any>) {
    return new DefaultControlInstance(sendMessage)
  }

  serialize(): [SerializedRecord, Transferable[]] {
    return serialize(this.config, {
      type: Definition.type,
      version: this.version,
    })
  }

  accept<R>(visitor: ControlDefinitionVisitor<R>, ...args: unknown[]): R {
    return visitor.visitColor(this, ...args)
  }
}

export class ColorDefinition<C extends Config = Config> extends Definition<C> {}

type UserConfig<D extends Config['defaultValue']> = Config & {
  defaultValue?: D
}

type NormedConfig<D extends Config['defaultValue']> = z.infer<
  SchemaByDefaultValue<D>['config']
>

export function Color<D extends Config['defaultValue']>(
  config?: UserConfig<D>,
): ColorDefinition<NormedConfig<D>> {
  return new ColorDefinition<NormedConfig<D>>(config ?? {}, 1)
}
