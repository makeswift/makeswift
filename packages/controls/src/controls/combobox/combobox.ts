import { z } from 'zod'

import { safeParse, type ParseResult } from '../../lib/zod'

import { Schema, type Data } from '../../common'
import { type CopyContext } from '../../context'
import {
  type DeserializedRecord,
  type SerializedRecord,
} from '../../serialization'

import {
  ControlDefinition,
  serialize,
  type Resolvable,
  type SchemaType,
} from '../definition'
import { DefaultControlInstance, type SendMessage } from '../instance'

type Option<T extends Data> = { id: string; value: T; label: string }
type GetOptionsType<T extends Data> = (
  query: string,
) => Option<T>[] | Promise<Option<T>[]>

type Config<T extends Data = Data> = {
  label?: string
  description?: string
  getOptions: GetOptionsType<T>
}

type ItemType<C extends Config> = C extends Config<infer Item> ? Item : never
type DataType<C extends Config> = Option<ItemType<C>>
type ValueType<C extends Config> = DataType<C>
type ResolvedValueType<C extends Config> =
  | Option<ItemType<C>>['value']
  | undefined

class Definition<C extends Config> extends ControlDefinition<
  typeof Definition.type,
  C,
  DataType<C>,
  ValueType<C>,
  ResolvedValueType<C>
> {
  static readonly type = 'makeswift::controls::combobox' as const

  static schema<T extends Data>(item: SchemaType<T>) {
    const type = z.literal(Definition.type)

    const data = z.object({
      id: z.string(),
      value: item,
      label: z.string(),
    }) as SchemaType<Option<T>>

    const value = data
    const resolvedValue = item.optional()

    const options = z.array(data)

    const config = z.object({
      getOptions: z
        .function()
        .args(z.string())
        .returns(z.union([options, z.promise(options)])),
      label: z.string().optional(),
      description: z.string().optional(),
    })

    const definition = z.object({
      type,
      config,
    })

    return {
      type,
      data,
      value,
      resolvedValue,
      config,
      definition,
    }
  }

  static deserialize(data: DeserializedRecord): ComboboxDefinition {
    if (data.type !== Definition.type) {
      throw new Error(
        `Combobox: expected type ${Definition.type}, got ${data.type}`,
      )
    }

    const { config } = Definition.schema(Schema.data).definition.parse(data)

    return Combobox(config)
  }

  get controlType() {
    return Definition.type
  }

  get schema() {
    return Definition.schema(Schema.data as SchemaType<ItemType<C>>)
  }

  safeParse(data: unknown | undefined): ParseResult<DataType<C> | undefined> {
    return safeParse(this.schema.data, data)
  }

  fromData(data: DataType<C> | undefined): ValueType<C> | undefined {
    return data
  }

  toData(value: ValueType<C>): DataType<C> {
    return value
  }

  copyData(
    data: DataType<C> | undefined,
    _context: CopyContext,
  ): DataType<C> | undefined {
    return data
  }

  resolveValue(
    data: DataType<C> | undefined,
  ): Resolvable<ResolvedValueType<C> | undefined> {
    return {
      name: Definition.type,
      readStable: () => this.fromData(data)?.value,
      subscribe: () => () => {},
      triggerResolve: async () => {},
    }
  }

  createInstance(sendMessage: SendMessage<any>) {
    return new DefaultControlInstance(sendMessage)
  }

  serialize(): [SerializedRecord, Transferable[]] {
    return serialize(this.config, {
      type: Definition.type,
    })
  }
}

export class ComboboxDefinition<
  C extends Config = Config,
> extends Definition<C> {}

type NormedConfig<
  T extends Data,
  GetOptions extends GetOptionsType<T>,
> = Config<T> & {
  getOptions: GetOptions
}

export function Combobox<T extends Data, GetOptions extends GetOptionsType<T>>(
  config: Config<T> & { getOptions: GetOptions },
): ComboboxDefinition<NormedConfig<T, GetOptions>> {
  return new ComboboxDefinition(config)
}
