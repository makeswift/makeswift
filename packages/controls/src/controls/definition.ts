import { z } from 'zod'

import { type ValueSubscription } from '../lib/value-subscription'
import { type ParseResult } from '../lib/zod'

import { type Data } from '../common/types'
import {
  type CopyContext,
  type MergeContext,
  type MergeTranslatableDataContext,
} from '../context'
import { type IntrospectionTarget } from '../introspection'
import { type ResourceResolver } from '../resources/resolver'
import {
  Serializable,
  serializeObject,
  type SerializedRecord,
} from '../serialization'
import { type Stylesheet } from '../stylesheet'

import { ControlInstance, type SendMessage } from './instance'

export type SchemaType<T> = z.ZodType<T>
export type SchemaTypeAny = SchemaType<any> | z.ZodBranded<SchemaType<any>, any>

export type Resolvable<T> = ValueSubscription<T> & {
  triggerResolve(currentValue?: T): Promise<unknown>
}

export abstract class ControlDefinition<
  ControlType extends string = string,
  Config = unknown,
  DataType extends Data = Data,
  ValueType extends Data = Data,
  ResolvedValueType = Data | unknown,
  InstanceType extends ControlInstance<any> = ControlInstance<any>,
> extends Serializable {
  // workaround for TypeScript type inference issues: https://bit.ly/4g2RvOQ
  __associatedTypes(_: {
    ControlType: ControlType
    Config: Config
    DataType: DataType
    ValueType: ValueType
    ResolvedValueType: ResolvedValueType
    InstanceType: InstanceType
  }) {}

  constructor(readonly config: Config) {
    super()
  }

  abstract get controlType(): ControlType

  abstract get schema(): {
    definition: SchemaType<unknown>
    type: SchemaType<ControlType>
    data: SchemaType<DataType>
    value: SchemaType<ValueType>
    resolvedValue: SchemaType<ResolvedValueType>
  }

  abstract safeParse(
    data: unknown | undefined,
  ): ParseResult<DataType | undefined>

  abstract fromData(data: DataType | undefined): ValueType | undefined
  abstract toData(value: ValueType): DataType

  abstract copyData(
    data: DataType | undefined,
    context: CopyContext,
  ): DataType | undefined

  mergeData(
    base: DataType,
    override: DataType | undefined,
    _context: MergeContext,
  ): DataType {
    return override ?? base
  }

  getTranslatableData(_data: DataType): Data {
    return null
  }

  mergeTranslatedData(
    data: DataType,
    _translatedData: Data,
    _context: MergeTranslatableDataContext,
  ): Data {
    return data as Data
  }

  abstract resolveValue(
    data: DataType | undefined,
    resolver: ResourceResolver,
    stylesheet: Stylesheet,
    control?: InstanceType,
  ): Resolvable<ResolvedValueType | undefined>

  abstract createInstance(sendMessage: SendMessage<any>): InstanceType

  abstract serialize(): [SerializedRecord, Transferable[]]

  introspect<R>(
    data: DataType | undefined,
    target: IntrospectionTarget<R>,
  ): R[] {
    return target.introspect(data)
  }
}

export function serialize(
  config: unknown,
  rest: { type: string } & Record<string, unknown>,
): [SerializedRecord, Transferable[]] {
  const [serializedConfig, transferables] = serializeObject(config)
  return [
    {
      config: serializedConfig,
      ...rest,
    } as unknown as SerializedRecord,
    transferables,
  ]
}
