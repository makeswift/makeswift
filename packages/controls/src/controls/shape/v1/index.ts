export * from './shape'
export * from './shape-control'
