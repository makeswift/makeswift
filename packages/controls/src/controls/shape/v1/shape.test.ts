import { createReplacementContext } from '../../../context'
import { Targets } from '../../../introspection'
import { deserializeRecord, type DeserializedRecord } from '../../../serialization'

import { type DataType, type ValueType } from '../../associated-types'
import { Checkbox, CheckboxDefinition } from '../../checkbox'
import { Color, ColorDefinition } from '../../color'
import { Combobox } from '../../combobox'
import { ControlDefinition } from '../../definition'
import { Image } from '../../image'
import { GenericLink as Link } from '../../link'
import { List } from '../../list'
import { Number } from '../../number'
import { TextArea } from '../../text-area'
import { TextInput } from '../../text-input'

import { Shape, ShapeDefinition } from './shape'

function deserializer(record: DeserializedRecord): ControlDefinition {
  switch (record.type) {
    case ShapeDefinition.type:
      return ShapeDefinition.deserialize(record, deserializer)

    case ColorDefinition.type:
      return ColorDefinition.deserialize(record)

    case CheckboxDefinition.type:
      return CheckboxDefinition.deserialize(record)

    default:
      throw new Error('Invalid control type')
  }
}

describe('Shape', () => {
  describe('constructor', () => {
    test('heterogenous controls', () => {
      const shape = Shape({
        type: {
          val: Color({ defaultValue: 'red' }),
          num: Checkbox(),
        },
      })
      expect(shape).toMatchSnapshot()
    })

    test('disallows extraneous properties', () => {
      Shape({
        type: {
          toggle: Checkbox(),
        },
        // @ts-expect-error
        extra: 'extra',
      })
    })
  })

  describe('assignability', () => {
    function assignTest(_def: ShapeDefinition) {}
    assignTest(
      Shape({
        type: {},
      }),
    )

    assignTest(
      Shape({
        type: {
          toggle: Checkbox(),
          color: Color({ defaultValue: 'red' }),
        },
      }),
    )

    assignTest(
      Shape({
        type: {
          combo: Combobox({
            getOptions: () => [{ id: '0', value: 'a', label: 'Alpha' }],
          }),
          link: Link({ label: 'Link' }),
        },
      }),
    )
  })

  describe('serialization', () => {
    test('serialize/deserialize shape of controls', () => {
      const shape = Shape({
        type: {
          color: Color({ defaultValue: 'red' }),
          checkbox: Checkbox(),
        },
      })

      const [serialized, _] = shape.serialize()
      expect(serialized).toMatchSnapshot()

      const deserialized = ShapeDefinition.deserialize(
        deserializeRecord(serialized),
        deserializer,
      )

      expect(deserialized).toEqual(shape)
    })

    test('serialize/deserialize composable shape of controls', () => {
      const shape = Shape({
        type: {
          shape: Shape({
            type: {
              color: Color({ defaultValue: 'red' }),
              checkbox: Checkbox(),
            },
          }),
        },
      })

      const [serialized, _] = shape.serialize()
      expect(serialized).toMatchSnapshot()

      const deserialized = ShapeDefinition.deserialize(
        deserializeRecord(serialized),
        deserializer,
      )

      expect(deserialized).toEqual(shape)
    })
  })

  describe('introspection', () => {
    test('is correctly delegated to subcontrols', () => {
      const shape = Shape({
        type: {
          color: Color({ defaultValue: 'red' }),
          link: Link(),
          image: Image(),
        },
      })

      const shapeData = {
        color: {
          swatchId: 'swatch-id',
          alpha: 1,
        },
        link: {
          type: 'OPEN_PAGE' as const,
          payload: {
            pageId: 'page-id',
            openInNewTab: false,
          },
        },
        image: 'file-id',
      }

      const swatchIds = shape.introspect(shapeData, Targets.Swatch)
      expect(swatchIds).toEqual(['swatch-id'])

      const pageIds = shape.introspect(shapeData, Targets.Page)
      expect(pageIds).toEqual(['page-id'])

      const fileIds = shape.introspect(shapeData, Targets.File)
      expect(fileIds).toEqual(['file-id'])
    })
  })

  describe('getTranslatableData', () => {
    test('is correctly delegated to subcontrols', () => {
      const shape = Shape({
        type: {
          textArea: TextArea(),
          textInput: TextInput(),
          color: Color({ defaultValue: 'red' }),
        },
      })

      const shapeData = {
        textArea: 'alpha',
        textInput: 'beta',
        color: {
          swatchId: 'swatch-id',
          alpha: 1,
        },
      }

      const translatableData = shape.getTranslatableData(shapeData)
      expect(translatableData).toEqual({
        textArea: 'alpha',
        textInput: 'beta',
        color: null,
      })
    })
  })

  describe('gracefully handles missing/extra props', () => {
    const shape = Shape({
      type: {
        color: Color({ defaultValue: 'red' }),
        list: List({ type: Number() }),
        link: Link(),
        text: TextArea(),
      },
    })

    const data = {
      // missing props + extra prop that should be ignored
      color: { swatchId: '[swatch-id]', alpha: 1 },
      text: 'Hello world',
      extraProp: 'extra',
    } as any as DataType<typeof shape>

    const value = {
      color: { swatchId: '[swatch-id]', alpha: 1 },
      list: [17],
      link: {
        type: 'OPEN_PAGE',
        payload: { pageId: '[page-id]', openInNewTab: false },
      },
      text: 'Hello world',
      // extra prop only, `toData` doesn't handle missing props
      extraProp: 'extra',
    } as ValueType<typeof shape>

    test('fromData', () => expect(shape.fromData(data)).toMatchSnapshot())
    test('toData', () => expect(shape.toData(value)).toMatchSnapshot())
    test('copyData', () =>
      expect(
        shape.copyData(data, {
          replacementContext: createReplacementContext({}),
          copyElement: (node) => node,
        }),
      ).toMatchSnapshot())

    test('getTranslatableData', () =>
      expect(shape.getTranslatableData(data)).toMatchSnapshot())

    test('mergeTranslatedData', () =>
      expect(
        shape.mergeTranslatedData(
          data,
          {},
          {
            translatedData: {},
            mergeTranslatedData: (node) => node,
          },
        ),
      ).toMatchSnapshot())

    test('introspect', () =>
      expect(shape.introspect(data, Targets.ChildrenElement)).toStrictEqual([]))
  })
})
