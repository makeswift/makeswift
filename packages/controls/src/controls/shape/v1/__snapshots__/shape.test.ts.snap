// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Shape constructor heterogenous controls 1`] = `
ShapeDefinition {
  "config": {
    "type": {
      "num": CheckboxDefinition {
        "config": {},
        "version": 1,
      },
      "val": ColorDefinition {
        "config": {
          "defaultValue": "red",
        },
        "version": 1,
      },
    },
  },
}
`;

exports[`Shape gracefully handles missing/extra props copyData 1`] = `
{
  "color": {
    "alpha": 1,
    "swatchId": "[swatch-id]",
  },
  "link": undefined,
  "list": undefined,
  "text": "Hello world",
}
`;

exports[`Shape gracefully handles missing/extra props fromData 1`] = `
{
  "color": {
    "alpha": 1,
    "swatchId": "[swatch-id]",
  },
  "link": undefined,
  "list": undefined,
  "text": "Hello world",
}
`;

exports[`Shape gracefully handles missing/extra props getTranslatableData 1`] = `
{
  "color": null,
  "link": null,
  "list": null,
  "text": "Hello world",
}
`;

exports[`Shape gracefully handles missing/extra props mergeTranslatedData 1`] = `
{
  "color": {
    "alpha": 1,
    "swatchId": "[swatch-id]",
  },
  "link": undefined,
  "list": undefined,
  "text": "Hello world",
}
`;

exports[`Shape gracefully handles missing/extra props toData 1`] = `
{
  "color": {
    "@@makeswift/type": "color::v1",
    "alpha": 1,
    "swatchId": "[swatch-id]",
  },
  "link": {
    "payload": {
      "openInNewTab": false,
      "pageId": "[page-id]",
    },
    "type": "OPEN_PAGE",
  },
  "list": [
    {
      "id": "xxxxxxxx-xxxx-xxxx-xxxx-100000000000",
      "type": "makeswift::controls::number",
      "value": {
        "@@makeswift/type": "number::v1",
        "value": 17,
      },
    },
  ],
  "text": {
    "@@makeswift/type": "text-area::v1",
    "value": "Hello world",
  },
}
`;

exports[`Shape serialization serialize/deserialize composable shape of controls 1`] = `
{
  "config": {
    "type": {
      "shape": {
        "config": {
          "type": {
            "checkbox": {
              "config": {},
              "type": "makeswift::controls::checkbox",
              "version": 1,
            },
            "color": {
              "config": {
                "defaultValue": "red",
              },
              "type": "makeswift::controls::color",
              "version": 1,
            },
          },
        },
        "type": "makeswift::controls::shape",
      },
    },
  },
  "type": "makeswift::controls::shape",
}
`;

exports[`Shape serialization serialize/deserialize shape of controls 1`] = `
{
  "config": {
    "type": {
      "checkbox": {
        "config": {},
        "type": "makeswift::controls::checkbox",
        "version": 1,
      },
      "color": {
        "config": {
          "defaultValue": "red",
        },
        "type": "makeswift::controls::color",
        "version": 1,
      },
    },
  },
  "type": "makeswift::controls::shape",
}
`;
